lib/teleterm/services/desktop/desktop.go

```diff
@@ -239,16 +311,22 @@ func (d *serverInterceptor) handleSharedDirectoryListRequest(r tdp.SharedDirecto
 }
 
 func (d *serverInterceptor) handleSharedDirectoryReadRequest(r tdp.SharedDirectoryReadRequest, sendToServer func(message tdp.Message) error) error {
-	contents, err := d.directoryAccess.Read(r.Path, int64(r.Offset), r.Length)
+	dirAccess, err := d.directoryAccessProvider.GetDirectoryAccess()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	buf := make([]byte, r.Length)
+	n, err := dirAccess.Read(r.Path, int64(r.Offset), buf)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	err = sendToServer(tdp.SharedDirectoryReadResponse{
 		CompletionID:   r.CompletionID,
 		ErrCode:        uint32(SharedDirectoryErrCodeNil),
-		ReadDataLength: uint32(len(contents)),
-		ReadData:       contents,
+		ReadDataLength: uint32(n),
+		ReadData:       buf[:n],
```

Comment:
Is there an equivalent of this code in the Web UI so that I can compare implementation? I guess there isn't since in the Web UI it's done by the browser API, right?

It's just that I don't have good heuristics for reviewing code that works on buffers, so I worry that I might miss something. ;f

Reply:
The code for Web UI is here, but it uses File System API:  https://github.com/gravitational/teleport/blob/d3fbeeb418b8eec99c62c540136f65568ca15a35/web/packages/shared/libs/tdp/sharedDirectoryAccess.ts#L171